plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -0,0 +1,699 @@
+/*
+ * Copyright (C) 2020 Aleksander Morgado <aleksander@aleksander.es>
+ * Copyright (C) 2021 Quectel Wireless Solutions Co., Ltd.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <sys/ioctl.h>
+
+#include "fu-firehose-updater.h"
+
+/* Maximum amount of non-"response" (e.g. "log") XML messages that can be
+ * received from the module when expecting a "response". This is just a safe
+ * upper limit to avoid reading forever. */
+#define MAX_RECV_MESSAGES 100
+
+/* When initializing the conversation with the firehose interpreter, the
+ * first step is to receive and process a bunch of messages sent by the
+ * module. The initial timeout to receive the first message is longer in case
+ * the module needs some initialization time itself; all the messages after
+ * the first one are expected to be received much quicker. The default timeout
+ * value should not be extremely long because the initialization phase ends
+ * when we don't receive more messages, so it's expected that the timeout will
+ * fully elapse after the last message sent by the module. */
+#define INITIALIZE_INITIAL_TIMEOUT_MS 3000
+#define INITIALIZE_TIMEOUT_MS         250
+
+/* Maximum amount of time to wait for a message from the module. */
+#define DEFAULT_RECV_TIMEOUT_MS 15000
+
+/* The first configure attempt sent to the module will include all the defaults
+ * listed below. If the module replies with a NAK specifying a different
+ * (shorter) max payload size to use, the second configure attempt will be done
+ * with that new suggested max payload size value. Only 2 configure attempts are
+ * therefore expected. */
+#define MAX_CONFIGURE_ATTEMPTS 2
+
+/* Defaults for the firehose configuration step. The max payload size to target
+ * in bytes may end up being a different if the module requests a shorter one.
+ */
+#define CONFIGURE_MEMORY_NAME                                   "nand"
+#define CONFIGURE_VERBOSE                                       0
+#define CONFIGURE_ALWAYS_VALIDATE                               0
+#define CONFIGURE_MAX_DIGEST_TABLE_SIZE_IN_BYTES                2048
+#define CONFIGURE_MAX_PAYLOAD_SIZE_TO_TARGET_IN_BYTES           8192
+#define CONFIGURE_ZLP_AWARE_HOST                                1
+#define CONFIGURE_SKIP_STORAGE_INIT                             0
+
+enum {
+	SIGNAL_WRITE_PROGRESS,
+	SIGNAL_LAST
+};
+
+static guint signals [SIGNAL_LAST] = { 0 };
+
+struct _FuFirehoseUpdater {
+	GObject		 parent_instance;
+	gchar		*port;
+	FuIOChannel	*io_channel;
+};
+
+G_DEFINE_TYPE (FuFirehoseUpdater, fu_firehose_updater, G_TYPE_OBJECT)
+
+static void
+fu_firehose_updater_log_message (const gchar *action, GBytes *msg)
+{
+	const gchar *msg_data;
+	gsize msg_size;
+
+	if (g_getenv ("FWUPD_MODEM_MANAGER_VERBOSE") == NULL)
+		return;
+
+	msg_data = (const gchar *) g_bytes_get_data (msg, &msg_size);
+	if (msg_size > G_MAXINT)
+		return;
+
+	g_debug ("%s:\n%.*s", action, (gint) msg_size, msg_data);
```

Comment:
are we sure msg_data is a string? Would fu_common_strsafe be a good idea?

Reply:
it is only used to print xml string, but I have added fu_common_strsafe just to be safe